tracing/ops_tracing/buffer.py

```diff
@@ -0,0 +1,298 @@
+                    -- observed events are more important
+                    priority INTEGER NOT NULL,
+                    -- a chunk of tracing data
+                    data BLOB NOT NULL,
+                    -- MIME type for these data
+                    mime TEXT NOT NULL
+                )
+                """
+            )
+            conn.execute(
+                """
+                CREATE INDEX IF NOT EXISTS tracing_priority_id
+                ON tracing
+                (priority, id)
+                """
+            )
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS settings (
+                    -- setting key, like "url" and "ca"
+                    key TEXT PRIMARY KEY,
+                    -- the value, with None coerced to empty string
+                    value TEXT NOT NULL
+                )
+                """
+            )
+
+    @contextlib.contextmanager
+    def tx(self, *, timeout: float = DB_TIMEOUT, readonly: bool = False):
+        """Thread-safe transaction context manager."""
+        with sqlite3.connect(self.path, isolation_level=None, timeout=timeout) as conn:
+            mode = 'DEFERRED' if readonly else 'IMMEDIATE'
+            conn.execute(f'BEGIN {mode}')
+            try:
+                yield conn
+            except:
+                conn.execute('ROLLBACK')
+                raise
+            else:
+                conn.execute('COMMIT')
+
+    @retry
+    def get_destination(self) -> Config:
+        """Get the tracing destination from the database."""
+        with self.tx(readonly=True) as conn:
+            settings = {k: v for k, v in conn.execute("""SELECT key, value FROM settings""")}
+            return Config(settings.get('url') or None, settings.get('ca') or None)
+
+    @retry
+    def set_destination(self, config: Config) -> None:
+        """Update the tracing destination in the database."""
+        with self.tx() as conn:
+            conn.execute(
+                """REPLACE INTO settings(key, value) VALUES('url', ?)""", (config.url or '',)
+            )
+            conn.execute(
+                """REPLACE INTO settings(key, value) VALUES('ca', ?)""", (config.ca or '',)
+            )
+
+    @retry
+    def mark_observed(self):
+        """Mark the tracing data collected in this dispatch as higher priority."""
+        if self.observed:
+            return
+
+        with self.tx(timeout=LONG_DB_TIMEOUT) as conn:
+            conn.execute(
+                f"""
+                UPDATE tracing
+                SET priority = ?
+                WHERE id IN ({','.join(('?',) * len(self.ids))})
+                """,  # noqa: S608
+                (OBSERVED_PRIORITY, *tuple(self.ids)),
+            )
+        self.observed = True
+        self.ids.clear()
+
+    @retry
+    def pump(self, record: tuple[bytes, str] | None = None) -> tuple[int, bytes, str] | None:
+        """Pump the buffer queue.
```

Comment:
I don't really understand the word or name "pump", and the docstring uses the word in its definition. :-) Can we come up with a better docstring at least, and perhaps a better name?

Reply:
I guess the analogy I used was flawed...

I was thinking "I push this much into the queue, and that much comes out (same or different record).", like pumping water into a hose.

I can't think of a good name off the top of my head, so I'm very open to suggestions!

Should I call it `pushpop`, like `heapq.heappushpop`?

Reply:
`pushpop` or `pushpop` seems okay. However, I actually think the name `pump` would be okay if the docstring sentence said something a bit more descriptive, "Push the given record into the queue, and pop the oldest one out" (or whatever is accurate).